File: typescript/packages/extensions/src/builder-code/cbor.ts
/**
 * ERC-8021 Schema 2 CBOR encoding for builder code suffixes.
 *
 * Schema 2 suffix format:
 *   [cbor_data (variable)] [suffix_data_length (2 bytes)] [schema_id = 0x02 (1 byte)] [ERC-8021 marker (16 bytes)]
 *
 * CBOR payload uses single-letter keys:
 *   "a" — app builder code (string)
 *   "w" — wallet/facilitator builder code (string)
 *   "s" — service codes (string array)
 */

import { type Hex } from "viem";
import { ERC_8021_MARKER, SCHEMA_2_ID, type BuilderCodeExtensionData } from "./types";

/**
 * Normalizes the `s` field (string or array of strings) into an array of strings.
 *
 * @param s - Service code value as a string, array of strings, or undefined
 * @returns Array of service code strings (empty when absent)
 */
function normalizeServiceCodes(s: BuilderCodeExtensionData["s"]): string[] {
  if (typeof s === "string") return [s];
  if (Array.isArray(s)) return s;
  return [];
}

/**
 * Encodes a CBOR map from builder code extension data.
 *
 * Produces a minimal CBOR map with:
 * - "a" key (major type 3, length 1) → string value (app/service code)
 * - "w" key (major type 3, length 1) → string value (facilitator code)
 * - "s" key (major type 3, length 1) → array of strings (related services)
 *
 * Uses hand-rolled CBOR to avoid adding a dependency.
 *
 * @param data - Builder code extension fields to encode
 * @returns CBOR-encoded map bytes
 */
function encodeCborMap(data: BuilderCodeExtensionData): Uint8Array {
  const entries: Uint8Array[] = [];
  let mapSize = 0;

  if (data.a) {
    mapSize++;
    entries.push(encodeCborString("a"));
    entries.push(encodeCborString(data.a));
  }

  if (data.w) {
    mapSize++;
    entries.push(encodeCborString("w"));
    entries.push(encodeCborString(data.w));
  }

  const serviceCodes = normalizeServiceCodes(data.s);
  if (serviceCodes.length > 0) {
    mapSize++;
    entries.push(encodeCborString("s"));
    entries.push(encodeCborArray(serviceCodes));
  }

  // CBOR map header
  const header = encodeCborMajorType(5, mapSize); // major type 5 = map

  const totalLength = header.length + entries.reduce((sum, e) => sum + e.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;

  result.set(header, offset);
  offset += header.length;

  for (const entry of entries) {
    result.set(entry, offset);
    offset += entry.length;
  }

  return result;
}

/**
 * Encodes a CBOR text string (major type 3).
 *
 * @param value - UTF-8 text to encode
 * @returns CBOR-encoded text string bytes
 */
function encodeCborString(value: string): Uint8Array {
  const encoded = new TextEncoder().encode(value);
  const header = encodeCborMajorType(3, encoded.length); // major type 3 = text string
  const result = new Uint8Array(header.length + encoded.length);
  result.set(header, 0);
  result.set(encoded, header.length);
  return result;
}

/**
 * Encodes a CBOR array of strings (major type 4).
 *
 * @param values - UTF-8 strings to encode as array elements
 * @returns CBOR-encoded array bytes
 */
function encodeCborArray(values: string[]): Uint8Array {
  const header = encodeCborMajorType(4, values.length); // major type 4 = array
  const encodedValues = values.map(encodeCborString);

  const totalLength = header.length + encodedValues.reduce((sum, e) => sum + e.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;

  result.set(header, offset);
  offset += header.length;

  for (const encoded of encodedValues) {
    result.set(encoded, offset);
    offset += encoded.length;
  }

  return result;
}

/**
 * Encodes a CBOR major type with an argument value.
 *
 * CBOR encoding rules:
 * - 0-23: single byte (major type << 5 | value)
 * - 24-255: two bytes (major type << 5 | 24, value)
 * - 256-65535: three bytes (major type << 5 | 25, value high, value low)
 *
 * @param majorType - CBOR major type (0–7)
 * @param value - Argument length or inline value for the header
 * @returns CBOR header bytes
 */
function encodeCborMajorType(majorType: number, value: number): Uint8Array {
  const mt = majorType << 5;

  if (value <= 23) {
    return new Uint8Array([mt | value]);
  }
  if (value <= 0xff) {
    return new Uint8Array([mt | 24, value]);
  }
  if (value <= 0xffff) {
    return new Uint8Array([mt | 25, (value >> 8) & 0xff, value & 0xff]);
  }

  throw new Error(`CBOR value too large: ${value}`);
}

/**
 * Builds a complete ERC-8021 Schema 2 data suffix from builder code data.
 *
 * Format: [cbor_data][suffix_data_length (2 bytes)][schema_id (1 byte)][marker (16 bytes)]
 *
 * The suffix_data_length covers the cbor_data only (not itself, schema_id, or marker).
 *
 * @param data - Builder code extension data with "a" and/or "s" fields
 * @returns Hex-encoded suffix bytes (without 0x prefix) ready to append to calldata
 */
export function encodeBuilderCodeSuffix(data: BuilderCodeExtensionData): Hex {
  const cborBytes = encodeCborMap(data);
  const cborLength = cborBytes.length;

  // suffix_data_length is 2 bytes, big-endian
  const lengthHigh = (cborLength >> 8) & 0xff;
  const lengthLow = cborLength & 0xff;

  // Build the full suffix: [cbor][length 2B][schema_id 1B][marker 16B]
  const suffixBytes = new Uint8Array(cborLength + 2 + 1 + 16);
  let offset = 0;

  // CBOR data
  suffixBytes.set(cborBytes, offset);
  offset += cborLength;

  // Suffix data length (2 bytes, big-endian)
  suffixBytes[offset++] = lengthHigh;
  suffixBytes[offset++] = lengthLow;

  // Schema ID
  suffixBytes[offset++] = SCHEMA_2_ID;

  // ERC-8021 marker (16 bytes)
  const markerBytes = hexToBytes(ERC_8021_MARKER);
  suffixBytes.set(markerBytes, offset);

  return `0x${bytesToHex(suffixBytes)}`;
}

/**
 * Converts a hex string (without 0x prefix) to bytes.
 *
 * @param hex - Hex-encoded string (two characters per byte)
 * @returns Decoded byte array
 */
function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

/**
 * Converts bytes to a lowercase hex string (without 0x prefix).
 *
 * @param bytes - Raw bytes to encode
 * @returns Lowercase hex string
 */
function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map(b => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Parses ERC-8021 Schema 2 builder code attribution from settlement calldata.
 *
 * @param calldata - Full transaction input data
 * @returns Decoded builder code fields, or undefined if no valid suffix is present
 */
export function parseBuilderCodeSuffixFromCalldata(
  calldata: Hex,
): BuilderCodeExtensionData | undefined {
  const hex = calldata.startsWith("0x") ? calldata.slice(2) : calldata;
  const markerPos = hex.lastIndexOf(ERC_8021_MARKER.toLowerCase());
  if (markerPos < 6) {
    return undefined;
  }

  if (parseInt(hex.slice(markerPos - 2, markerPos), 16) !== SCHEMA_2_ID) {
    return undefined;
  }

  const cborLength = parseInt(hex.slice(markerPos - 6, markerPos - 2), 16);
  const suffixStart = markerPos - 6 - cborLength * 2;
  if (suffixStart < 0 || suffixStart + (cborLength + 19) * 2 !== hex.length) {
    return undefined;
  }

  const bytes = hexToBytes(hex.slice(suffixStart, markerPos - 6));
  let o = 0;

  if (bytes[o] >> 5 !== 5) {
    return undefined;
  }

  const mapInfo = bytes[o++] & 0x1f;
  const mapSize = mapInfo <= 23 ? mapInfo : mapInfo === 24 ? bytes[o++] : undefined;
  if (mapSize === undefined) {
    return undefined;
  }

  const result: BuilderCodeExtensionData = {};
  for (let entry = 0; entry < mapSize; entry++) {
    if (bytes[o] >> 5 !== 3) {
      return undefined;
    }

    const keyInfo = bytes[o++] & 0x1f;
    const keyLen = keyInfo <= 23 ? keyInfo : keyInfo === 24 ? bytes[o++] : undefined;
    if (keyLen === undefined) {
      return undefined;
    }

    const key = new TextDecoder().decode(bytes.subarray(o, o + keyLen));
    o += keyLen;

    if (key === "a" || key === "w") {
      if (bytes[o] >> 5 !== 3) {
        return undefined;
      }

      const valueInfo = bytes[o++] & 0x1f;
      const valueLen = valueInfo <= 23 ? valueInfo : valueInfo === 24 ? bytes[o++] : undefined;
      if (valueLen === undefined) {
        return undefined;
      }

      result[key] = new TextDecoder().decode(bytes.subarray(o, o + valueLen));
      o += valueLen;
      continue;
    }

    if (key === "s") {
      if (bytes[o] >> 5 !== 4) {
        return undefined;
      }

      const arrayInfo = bytes[o++] & 0x1f;
      const arraySize = arrayInfo <= 23 ? arrayInfo : arrayInfo === 24 ? bytes[o++] : undefined;
      if (arraySize === undefined) {
        return undefined;
      }

      const codes: string[] = [];
      for (let i = 0; i < arraySize; i++) {
        if (bytes[o] >> 5 !== 3) {
          return undefined;
        }

        const itemInfo = bytes[o++] & 0x1f;
        const itemLen = itemInfo <= 23 ? itemInfo : itemInfo === 24 ? bytes[o++] : undefined;
        if (itemLen === undefined) {
          return undefined;
        }

        codes.push(new TextDecoder().decode(bytes.subarray(o, o + itemLen)));
        o += itemLen;
      }

      if (codes.length > 0) {
        result.s = codes;
      }
      continue;
    }

    return undefined;
  }

  return result;
}

File: python/x402/extensions/builder_code/cbor.py
"""ERC-8021 Schema 2 CBOR encoding for builder code suffixes.

Schema 2 suffix format::

    [cbor_data (variable)] [suffix_data_length (2 bytes)] [schema_id = 0x02 (1 byte)] [ERC-8021 marker (16 bytes)]

The CBOR payload uses single-letter keys:
- ``a`` — app builder code (string)
- ``w`` — wallet/facilitator builder code (string)
- ``s`` — service codes (string array)

Hand-rolled CBOR keeps the extension dependency-free (stdlib only).
"""

from __future__ import annotations

from .types import ERC_8021_MARKER, SCHEMA_2_ID, BuilderCodeExtensionData

# CBOR major types used by this encoder.
_MAJOR_TEXT_STRING = 3
_MAJOR_ARRAY = 4
_MAJOR_MAP = 5


def _normalize_service_codes(s: str | list[str] | None) -> list[str]:
    """Normalize the ``s`` field (string or list of strings) into a list."""
    if isinstance(s, str):
        return [s]
    if isinstance(s, list):
        return s
    return []


def _encode_major_type(major_type: int, value: int) -> bytes:
    """Encode a CBOR major type with its argument value.

    Rules:
    - 0-23: single byte ``(major_type << 5) | value``
    - 24-255: two bytes ``(major_type << 5) | 24``, value
    - 256-65535: three bytes ``(major_type << 5) | 25``, value (big-endian)
    """
    mt = major_type << 5
    if value <= 23:
        return bytes([mt | value])
    if value <= 0xFF:
        return bytes([mt | 24, value])
    if value <= 0xFFFF:
        return bytes([mt | 25, (value >> 8) & 0xFF, value & 0xFF])
    raise ValueError(f"CBOR value too large: {value}")


def _encode_string(value: str) -> bytes:
    """Encode a CBOR text string (major type 3)."""
    encoded = value.encode("utf-8")
    return _encode_major_type(_MAJOR_TEXT_STRING, len(encoded)) + encoded


def _encode_array(values: list[str]) -> bytes:
    """Encode a CBOR array of text strings (major type 4)."""
    result = _encode_major_type(_MAJOR_ARRAY, len(values))
    for value in values:
        result += _encode_string(value)
    return result


def _encode_cbor_map(data: BuilderCodeExtensionData) -> bytes:
    """Encode a minimal CBOR map from builder code data, in ``a``, ``w``, ``s`` order."""
    entries = bytearray()
    map_size = 0

    if data.a:
        map_size += 1
        entries += _encode_string("a")
        entries += _encode_string(data.a)

    if data.w:
        map_size += 1
        entries += _encode_string("w")
        entries += _encode_string(data.w)

    service_codes = _normalize_service_codes(data.s)
    if service_codes:
        map_size += 1
        entries += _encode_string("s")
        entries += _encode_array(service_codes)

    return _encode_major_type(_MAJOR_MAP, map_size) + bytes(entries)


def encode_builder_code_suffix(data: BuilderCodeExtensionData) -> str:
    """Build a complete ERC-8021 Schema 2 data suffix from builder code data.

    Format: ``[cbor_data][suffix_data_length (2 bytes)][schema_id (1 byte)][marker (16 bytes)]``.
    ``suffix_data_length`` covers the CBOR data only.

    Args:
        data: Builder code fields to encode.

    Returns:
        Hex-encoded suffix (with ``0x`` prefix) ready to append to calldata.
    """
    cbor_bytes = _encode_cbor_map(data)
    cbor_length = len(cbor_bytes)

    suffix = (
        cbor_bytes
        + bytes([(cbor_length >> 8) & 0xFF, cbor_length & 0xFF])
        + bytes([SCHEMA_2_ID])
        + bytes.fromhex(ERC_8021_MARKER)
    )
    return "0x" + suffix.hex()


def _read_length(byte: int, data: bytes, offset: int) -> tuple[int | None, int]:
    """Read a CBOR argument that is either inline (<=23) or a single following byte (24)."""
    info = byte & 0x1F
    if info <= 23:
        return info, offset
    if info == 24:
        return data[offset], offset + 1
    return None, offset


def parse_builder_code_suffix_from_calldata(
    calldata: str,
) -> BuilderCodeExtensionData | None:
    """Parse ERC-8021 Schema 2 builder code attribution from settlement calldata.

    Args:
        calldata: Full transaction input data (with or without ``0x`` prefix).

    Returns:
        Decoded builder code fields, or ``None`` if no valid suffix is present.
    """
    hex_str = calldata[2:] if calldata.startswith("0x") else calldata
    marker = ERC_8021_MARKER.lower()
    marker_pos = hex_str.lower().rfind(marker)
    if marker_pos < 6:
        return None

    if int(hex_str[marker_pos - 2 : marker_pos], 16) != SCHEMA_2_ID:
        return None

    cbor_length = int(hex_str[marker_pos - 6 : marker_pos - 2], 16)
    suffix_start = marker_pos - 6 - cbor_length * 2
    if suffix_start < 0 or suffix_start + (cbor_length + 19) * 2 != len(hex_str):
        return None

    data = bytes.fromhex(hex_str[suffix_start : marker_pos - 6])
    offset = 0

    if data[offset] >> 5 != _MAJOR_MAP:
        return None

    map_size, offset = _read_length(data[offset], data, offset + 1)
    if map_size is None:
        return None

    result = BuilderCodeExtensionData()
    for _ in range(map_size):
        if data[offset] >> 5 != _MAJOR_TEXT_STRING:
            return None
        key_len, offset = _read_length(data[offset], data, offset + 1)
        if key_len is None:
            return None
        key = data[offset : offset + key_len].decode("utf-8")
        offset += key_len

        if key in ("a", "w"):
            if data[offset] >> 5 != _MAJOR_TEXT_STRING:
                return None
            value_len, offset = _read_length(data[offset], data, offset + 1)
            if value_len is None:
                return None
            value = data[offset : offset + value_len].decode("utf-8")
            offset += value_len
            setattr(result, key, value)
            continue

        if key == "s":
            if data[offset] >> 5 != _MAJOR_ARRAY:
                return None
            array_size, offset = _read_length(data[offset], data, offset + 1)
            if array_size is None:
                return None
            codes: list[str] = []
            for _ in range(array_size):
                if data[offset] >> 5 != _MAJOR_TEXT_STRING:
                    return None
                item_len, offset = _read_length(data[offset], data, offset + 1)
                if item_len is None:
                    return None
                codes.append(data[offset : offset + item_len].decode("utf-8"))
                offset += item_len
            if codes:
                result.s = codes
            continue

        return None

    return result
